nexus/db-queries/src/db/datastore/oximeter.rs

```diff
@@ -168,94 +167,29 @@ impl DataStore {
-        // _except_ incorrect orderings where the types still line up (e.g.,
-        // swapping two Uuid columns), which is not ideal but is hopefully good
-        // enough.
-        let producer_subquery = {
-            use db::schema::oximeter::dsl;
-
-            dsl::oximeter
-                .select((
-                    producer.id().into_sql::<sql_types::Uuid>(),
-                    producer
-                        .time_created()
-                        .into_sql::<sql_types::Timestamptz>(),
-                    producer
-                        .time_modified()
-                        .into_sql::<sql_types::Timestamptz>(),
-                    producer.kind.into_sql::<ProducerKindEnum>(),
-                    producer.ip.into_sql::<sql_types::Inet>(),
-                    producer.port.into_sql::<sql_types::Int4>(),
-                    producer.interval.into_sql::<sql_types::Float8>(),
-                    producer.oximeter_id.into_sql::<sql_types::Uuid>(),
-                ))
-                .filter(
-                    dsl::id
-                        .eq(producer.oximeter_id)
-                        .and(dsl::time_expunged.is_null()),
-                )
-        };
-
-        use db::schema::metric_producer::dsl;
-
-        // TODO: see https://github.com/oxidecomputer/omicron/issues/323
-        let n = diesel::insert_into(dsl::metric_producer)
-            .values(producer_subquery)
-            .on_conflict(dsl::id)
-            .do_update()
-            .set((
-                dsl::time_modified.eq(Utc::now()),
-                dsl::kind.eq(producer.kind),
-                dsl::ip.eq(producer.ip),
-                dsl::port.eq(producer.port),
-                dsl::interval.eq(producer.interval),
-            ))
-            .execute_async(&*self.pool_connection_authorized(opctx).await?)
+        producer: &internal::nexus::ProducerEndpoint,
+    ) -> Result<OximeterInfo, Error> {
+        match queries::oximeter::upsert_producer(producer)
+            .get_result_async(&*self.pool_connection_authorized(opctx).await?)
             .await
-            .map_err(|e| {
-                public_error_from_diesel(
-                    e,
-                    ErrorHandler::Conflict(
-                        ResourceType::MetricProducer,
-                        "Producer Endpoint",
-                    ),
-                )
-            })?;
-
-        // We expect `n` to basically always be 1 (1 row was inserted or
-        // updated). It can be 0 if `producer.oximeter_id` doesn't exist or has
-        // been expunged. It can never be 2 or greater because
-        // `producer_subquery` filters on finding an exact row for its Oximeter
-        // instance's ID.
-        match n {
-            0 => Err(Error::not_found_by_id(
-                ResourceType::Oximeter,
-                &producer.oximeter_id,
+        {
+            Ok(info) => Ok(info),
+            Err(DieselError::NotFound) => Err(Error::unavail(
```

Comment:
What actually generates this `NotFound` error? Since we have `COALESCE` in the query now, it seems like we might also need to catch an error trying to insert NULL into the `producer.oximeter_id` column, but I'm not sure.

Reply:
The `NotFound` error is because we're going through [get_result()](https://docs.rs/diesel/2.2.4/diesel/prelude/trait.RunQueryDsl.html#method.get_result), which expects to get a row and returns `NotFound` if it can't.

`COALESCE(NULL, NULL)` would select null, but I believe the second argument (the ID for the "randomly chosen") is always guaranteed to be non-NULL. The clause is:

```sql
chosen_oximeter AS (
  SELECT COALESCE(existing_oximeter.id, random_oximeter.id) AS oximeter_id
  FROM random_oximeter LEFT JOIN existing_oximeter ON true
)
```

Since we're selecting from `random_oximeter LEFT JOIN ...`, we only get a value _to_ coalesce if the `random_oximeter` clause returned a row.

Reply:
Sorry, I didn't follow through that whole thought. If `random_oximeter` doesn't return any rows (because there are no non-expunged oximeters), then `chosen_oximeter` will also return no rows, so the subsequent `INSERT INTO ... FROM chosen_oximeter ... RETURNING ...` will not do an insert and therefore not return anything, so the final `SELECT` will also return no rows, finally bubbling out to `get_result()` which turns the whole thing into a `NotFound`.